// express-ejs-layouts/lib/express-layouts.js
var contentPattern = '&&<>&&'

/**
 * 模板扩展名数组
 * @since 2013-04-15
 */
var exts = [];

function contentFor(contentName) {
  return contentPattern + contentName + contentPattern
}

function parseContents(locals) {
  var str = locals.body
  var regex = new RegExp('\n?' + contentPattern + '.+?' + contentPattern + '\n?', 'g')
  var split = str.split(regex)
  var matches = str.match(regex)

  locals.body = split[0]
  var i = 1;

  if (matches != null) {
    matches.forEach(function(match) {
      var name = match.split(contentPattern)[1]
      locals[name] = split[i];
      i++;
    });
  }
}

function parseScripts(locals) {
  var str = locals.body
  locals.script = ''
  
  var regex = /\<script(.|\n)*?\>(.|\n)*?\<\/script\>/g
  if (regex.test(str)) {
    locals.body = str.replace(regex, '');
    locals.script = str.match(regex).join('\n');
  }
}

module.exports = exports = function(req, res, next) {
  var render = res.render;
  
  res.render = function(view, options, fn) {
    
    var options = options || {};
    
    // support callback function as second arg
    if ('function' == typeof options) {
      fn = options, options = {};
    }
    
    var app = req.app
    var defaultLayout = app.get('layout')
    var specifiedLayout = options.layout
    
    if (specifiedLayout === false || ((specifiedLayout || defaultLayout) === false) ) {
      render.call(res, view, options, fn)
      return
    }

    var layout = specifiedLayout || defaultLayout
    if (layout === true || layout === undefined)
      layout = 'layout'
    
    options.contentFor = contentFor
    
    var self = this;
    render.call(res, view, options, function(err, str){ 
      
      if (err) {
        if (fn)
          return fn(err)
        else
          throw err
      }
      
      var locals = { body: str }
      for (var l in options) {
        if (options.hasOwnProperty(l)
          && l != 'layout'
          && l != 'contentFor')
          locals[l] = options[l];
      }

      if (options.extractScripts === true 
        || (options.extractScripts === undefined && app.get('layout extractScripts') === true))
        parseScripts(locals)
        
      parseContents(locals)
      
      render.call(self, layout, locals, fn);
    });
  };
  
  next();
};

module.exports.register = function(ext){
    if(!ext.match(/^\./)){
        ext = '.' + ext;
    }
    exts.push(ext);
};

/**
 * 对指定文件类型的模板进行layout渲染
 * @param req
 * @param res
 * @param next
 * @author huhai
 * @since 2013-04-15
 */
module.exports.layout = function(req, res, next){
    var render = res.render;
    res.render = function(view, options, fn){
        for(var i in exts){
            if(view.match('.+' + exts[i] + '$')){//仅对匹配的模板进行layout渲染
                console.log('layout proxy ' + view);
                layoutProxy.call(res, req, res, render, view, options, fn);
                break;
            }else{
                render.call(res, view, options, fn);
            }
        }
    };
    next();
};

/**
 *
 * @param req
 * @param res
 * @param render
 * @param view
 * @param options
 * @param fn
 * @author huhai
 * @since 2013-04-15
 */
function layoutProxy(req, res, render, view, options, fn) {

    var options = options || {};

    // support callback function as second arg
    if ('function' == typeof options) {
        fn = options, options = {};
    }

    var app = req.app
    var defaultLayout = app.get('layout')
    var specifiedLayout = options.layout

    if (specifiedLayout === false || ((specifiedLayout || defaultLayout) === false) ) {
        render.call(res, view, options, fn)
        return
    }

    var layout = specifiedLayout || defaultLayout
    if (layout === true || layout === undefined)
        layout = 'layout'

    options.contentFor = contentFor

    var self = this;
    render.call(res, view, options, function(err, str){

        if (err) {
            if (fn)
                return fn(err)
            else
                throw err
        }

        var locals = { body: str }
        for (var l in options) {
            if (options.hasOwnProperty(l)
                && l != 'layout'
                && l != 'contentFor')
                locals[l] = options[l];
        }

        if (options.extractScripts === true
            || (options.extractScripts === undefined && app.get('layout extractScripts') === true))
            parseScripts(locals)

        parseContents(locals)

        render.call(self, layout, locals, fn);
    });
};
